lib/spack/spack/test/concretize_requirements.py

```diff
@@ -297,3 +297,25 @@ def test_requirements_are_higher_priority_than_deprecation(concretize_scope, tes
     s1 = Spec("y").concretized()
     assert s1.satisfies("@2.3")
     assert s1.satisfies("%gcc")
+
+
+@pytest.mark.parametrize("spec_str,requirement_str", [("x", "%gcc"), ("x", "%clang")])
+def test_default_requirements_with_all(spec_str, requirement_str, concretize_scope, test_repo):
+    """Test that users can override a deprecated version with a requirement."""
+    if spack.config.get("config:concretizer") == "original":
+        pytest.skip("Original concretizer does not support configuration" " requirements")
+
+    # @2.3 is a deprecated versions. Ensure that any_of picks both constraints,
+    # since they are possible
```

Comment:
I don't think this comment applies here.

Also it might be worth testing the combination of `all:` and specific package constraints

Reply:
Done in db7dab1